cpp/src/neighbors/detail/cagra/cagra_build.cuh

```diff
@@ -162,11 +162,26 @@ void build_knn_graph(
   // search top (k + 1) neighbors
   //
 
-  const auto top_k          = node_degree + 1;
-  uint32_t gpu_top_k        = node_degree * pq.refinement_rate;
-  gpu_top_k                 = std::min<IdxT>(std::max(gpu_top_k, top_k), dataset.extent(0));
-  const auto num_queries    = dataset.extent(0);
-  const auto max_batch_size = 1024;
+  const auto top_k       = node_degree + 1;
+  uint32_t gpu_top_k     = node_degree * pq.refinement_rate;
+  gpu_top_k              = std::min<IdxT>(std::max(gpu_top_k, top_k), dataset.extent(0));
+  const auto num_queries = dataset.extent(0);
+  rmm::device_async_resource_ref workspace_mr = raft::resource::get_workspace_resource(res);
+
+  constexpr size_t kDefaultBatchSize = 1024;
+  constexpr size_t kMaxBatchSize     = 4096;  // No more perf beyond this
+  constexpr size_t kMinBatchSize     = 128;   // Too slow if smaller
+  // Heuristic: how much of the workspace we can spare for the queries.
+  // The rest is going to be used by ivf_pq::search
+  const auto workspace_queries_bytes = raft::resource::get_workspace_free_bytes(res) / 5;
+  auto max_batch_size =
+    std::min<size_t>(workspace_queries_bytes / sizeof(DataT) / dataset.extent(1), kMaxBatchSize);
+  // Heuristic: if the workspace is too small for a decent batch size, switch to use the large
+  // resource with a default batch size.
+  if (max_batch_size < kMinBatchSize) {
+    max_batch_size = kDefaultBatchSize;
+    workspace_mr   = raft::resource::get_large_workspace_resource(res);
+  }
```

Comment:
While the intention is good, I am not sure if adding more heuristics is the way to go. The code looks good to me, but it adds complexity, and will make it more difficult to reproduce future memory errors.

Since IVF-PQ would anyways adjust its internal batch size according to the available workspace size, couldn't we just fix here `max_batch_size = 4096`.

Reply:
The trade-off is that taking more memory here will limit the memory available to IVF-PQ internal batching.

If we fix the batch size here and there's not enough memory, we could end up with a few nasty situations depending on the size:
  a) (very large batch) OOM throws here 
  b) (large batch) OOM throws further in IVF-PQ search (not sure if this one is possible)
  c) (a bit too large batch) IVF-PQ internally selects very small batch and runs slowly
  d) (small batch) IVF-PQ is limited by the given batch size and runs slowly

So I don't see how we can easily go here without a heuristics.

Reply:
Assuming `max_batch_size = 4096`, we are talking about `4096*O(graph_degree) + O(dataset_dim)` memory allocated by CAGRA build. This is negligible compared to the internal memory usage of IVF-PQ. Therefore I would just leave the constant large batch here, and leave it to IVF-PQ to do any internal batching if necessary. Please correct me if I am missing something.

Reply:
Hmm, looking through the code one more time, to me it seems that `build_knn_graph` itself uses the workspace resource only to batch the dataset. This is the asymptotic:
  1. build_knn_graph (vec_batches) :
      `max_batch_size * O(dataset_dim)`
  2. ivf_pq::search():
      `max_batch_size * (O(dataset_dim) + O(max_samples) + O(k * n_probes))`

Normally we only have problems with the non-fused version of IVF-PQ and the large probe count, so that the max_batch_size * max_samples can reach gigabytes-scale. Then the heuristics for the `build_knn_graph` seems indeed unnecessary.

Reply:
I've updated the code to leave just one extra constant `kMinWorkspaceRatio` and greatly simplify the logic. @tfeher  please have a look.

Reply:
ok, thanks!